src/wallet/wallet2.cpp

```diff
@@ -13099,4 +13099,128 @@ uint64_t wallet2::get_bytes_received() const
 {
   return m_http_client.get_bytes_received();
 }
+//----------------------------------------------------------------------------------------------------
+void wallet2::configure(const std::function<bool(const char*, std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>>&)> &f, const password_container &pwd_container)
+{
+  std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>> v;
+
+  v = {{tr("Always prompt before sending a transaction"), m_always_confirm_transfers}};
+  if (!f(NULL, v)) return;
+  m_always_confirm_transfers = boost::get<bool>(v[0].second);
```

Comment:
This design is interesting, but still somewhat awkward. The original value is going into the variant, but is used only for its type, and then you have to copy the value back out again. This funkiness is seen  with `std::pair<bool, std::uint64_t>`, which is just some pseudo-type to differentiate between numerical and money types.

Formally this is actually a `std::function<bool(const std::string&)>` where the result value is "bound" to the function. I.e.:
```c++
v = {
  tr("Always prompt before sending a transaction"),
  std::bind(&parse_bool, _1, std::ref(m_always_confirm_transfers))
};
```

which allows for any `std::string->T` conversion/storage to take place, and separates conversion function from type (currently conversion function is linked to type, thus the `std::pair<bool, std::uint64_t>` stuff).

Reply:
I was thinking a GUI would probably want to display a widget type which matches the variant type, so a parser function does not really make sense in wallet2. For a bool, it'd have a "selected or not" widget for example.

Reply:
BTW, the value isn't used only for its type. In simplewallet, you can press ENTER and it stays at its current value.

Reply:
> I was thinking a GUI would probably want to display a widget type which matches the variant type, so a parser function does not really make sense in wallet2. For a bool, it'd have a "selected or not" widget for example.

I'm not sure that the GUI would use as-is because it would force a wizard-style UI instead of a configuration screen (options are thrown to the callback one "group" at a time). Maybe the wizard-style is preferable in this case, I don't know.

If special widgets are needed adding an `enum` alongside the bound function would do it. It would be more readable/auditable than `std::pair<bool, std::uint64_t>`  representing the money type. Which brings me to the next thing - if you bound functions + `enum`s really aren't your thing then considering `struct`s for some of these types so its more self-documenting. And `typedef`ing the large `boost::variant` type would help too.

The other disadvantage to this current approach is that the GUI has to re-define conversion behavior (unless the visitor `struct` is moved to some header). With an `enum` + bound function a UI can just throw any string at the conversion function and check the status `bool`. Unfortunately these functions don't give detailed failure codes, so they may want custom functions for some or all of them anyway.

> BTW, the value isn't used only for its type. In simplewallet, you can press ENTER and it stays at its current value.

This behavior would be unchanged when using bound functions. If the callback is not invoked, then the bound value never changes.

Reply:
Actually, I'm mixed on the `enum` + bound function. Its definitely cleaner in some ways, but the `boost::variant` is an enum + value in one. The bound function seems slightly preferable, but it might be the grossness of `std::pair<bool, std;:uint64_t>` stuck in my head.

Reply:
That sounds maybe like what I'm doing now:

std::vector<std::vector<std::tuple<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>, std::function<void(std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>>&)>>>> wallet2::configure(const password_container &pwd_container)
{
  std::vector<std::vector<std::tuple<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>, std::function<void(std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>>&)>>>> v;

#define SET(type, var) \
  [&](std::vector<std::pair<const char*, boost::variant<uint64_t, uint32_t, std::pair<bool, uint64_t>, std::pair<uint32_t, std::vector<std::pair<uint32_t, std::string>>>, bool, std::string>>> &v0) { \
    var = boost::get<type>(v0[0].second); \
    rewrite(m_keys_file, pwd_container.password()); \
  }

  v.push_back({{tr("Always prompt before sending a transaction"), m_always_confirm_transfers, SET(bool, m_always_confirm_transfers)}});

  v.push_back{{tr("Print ring members before sending a transaction"), m_print_ring_members, SET(bool, m_print_ring_members)}};

Reply:
Only to the `push_back` part, but not the remainder (primarily a switch to enum + bound function or something less crummy than arbitrary `std::pair<bool, std::uint64_t>` types).

Also, having some of these visitors in some header might be useful, if the GUI were truly going to use it at some point.